lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -2145,6 +2168,102 @@ def construct_error_details_map_grpc(gax_error)
       {}
     end
 
+    # Construct monitored resource locally for k8s resources.
+    def contruct_k8s_resource_locally(local_resource_id)
+      matched_regexp_group = nil
+      # Regex match the local_resource_id to make sure we only do
+      # this for 'k8s_container' and 'k8s_node'.
+      [@compiled_k8s_node_local_resource_id_regexp,
+       @compiled_k8s_container_local_resource_id_regexp].each do |id_regexp|
+        matched_regexp_group ||= id_regexp.match(local_resource_id)
+      end
+      case matched_regexp_group['resource_type']
+      when 'k8s_container'
+        @log.debug('Detected local_resource_id in k8s_container format.' \
+                   "Constructing resource locally with #{local_resource_id}.")
+        return construct_k8s_container_resource(matched_regexp_group)
+      when 'k8s_node'
+        @log.debug('Detected local_resource_id in k8s_node format.' \
+                   "Constructing resource locally with #{local_resource_id}.")
+        return construct_k8s_node_resource(matched_regexp_group)
+      end
+    end
+
+    # Construct the k8s_container monitored resource locally. In case of any
+    # exception, return nil.
+    def construct_k8s_container_resource(matched_regexp_group)
+      constructed_resource = Google::Apis::LoggingV2::MonitoredResource.new(
+        type: 'k8s_container',
+        labels: {
+          'namespace_name' => matched_regexp_group['namespace_name'],
+          'pod_name' => matched_regexp_group['pod_name'],
+          'container_name' => matched_regexp_group['container_name'],
+          'cluster_name' => retrieve_k8s_cluster_name,
+          'location' => retrieve_k8s_location
+        }
+      )
+      @log.debug('Constructed k8s_container resource locally:\n' \
+                 "#{constructed_resource.inspect}")
+      constructed_resource
+    rescue StandardError => e
+      @log.error 'Failed to construct "k8s_container" resource locally. Fall' \
+                 ' back to write logs to instance resource.', error: e
+      # Fall back to default instance resource.
```

Comment:
To be fair, I think we fall back to the default resource detection logic (`@resource`, a.k.a. `determine_agent_level_monitored_resource_via_legacy`). So in the case of Kubernetes, we'll fall back on a `gke_container` resource. Am I missing something?

Reply:
To fall back on a `gke_container` resource, it requires the `detect_subservice` config. The latest configmap will not have that. Do we want to keep it? I'm leaning against it.

Reply:
Doesn't `detect_subservice` default to `true`?

Reply:
Sorry, missed this one in my previous review. I think we should leave `detect_subservice` alone — it serves as a fallback for the current functionality, and if we put it in a config map, we'll have a harder time deprecating it later.